Blog Post - Animated Scroll Effects (#471)
ritterim/ritterim.github.io: _posts/2022-05-18-animated-scroll-effects.md

---
layout: post
title: "Animated Scroll Effects"
date: 2022-05-18 12:00:00
tags:
- CSS
- SCSS
- JavaScript
- UI/UX
- frontend
categories:
- development
twitter_text:
authors: Ryan Trimble
image: https://images.unsplash.com/photo-1627163439134-7a8c47e08208?ixlib=rb-1.2.1&ixid=MnwxMjA3fDB8MHxwaG90by1wYWdlfHx8fGVufDB8fHx8&auto=format&fit=crop&w=2532&q=80
image_url: https://unsplash.com/photos/R5A_YlcSJwA
image_credit: Shubham Dhage
---

> I’m a huge fan of animation on the web, especially when it comes to user interaction. A simple way to include such a thing on webpages is by triggering some sort of effect to occur when scrolling.

Reviewer: I can't put my finger on it but the second sentence reads weird to me. I know you don't want to say animation again but `such a thing` is making me feel like I should be drinking my tea with my pinky out while reading this. I also don't know that you need `to occur when`. could it just be:
```suggestion
I’m a huge fan of animation on the web, especially when it comes to user interaction. A simple way to include such a thing on webpages is by triggering some sort of effect when scrolling.
```

Elements fading in and out to tell a story. Background colors changing to differentiate page sections. Cards and content popping in at various points on the page. These types of things can enhance the user experience on your page.

> The [intersection observer API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) provides a vanilla JavaScript approach to adding scroll effects. by triggering functions when elements intersect certain points of the viewport. Intersection observer can be difficult to work with, so a library such as [Animate on scroll](https://michalsnik.github.io/aos/) may make things simpler. 

Reviewer 2: ```suggestion
The [intersection observer API](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) provides a vanilla JavaScript approach to adding scroll effects by triggering functions when elements intersect certain points of the viewport. Intersection observer can be difficult to work with, so a library such as [Animate on scroll](https://michalsnik.github.io/aos/) may make things simpler. 
```

My personal favorite library to animate with is the [GreenSock Animation Platform](https://greensock.com/). GSAP is useful for all sorts of animations, including scroll triggered effects. 

To show this off, let’s build a neat effect I came across on a website recently! The website included a hero section containing text. As you scroll the page, the text scales to highlight the importance of the message.

> For this demo I have created a [Codepen](https://codepen.io/mrtrimble/pen/gOvLPgM??editors=0010) to follow along with. I already included HTML, CSS, GSAP, and the ScrollTrigger plugin, so no need to setup anything! We’ll focus on the JavaScript to find out that with a few lines of code, we can build a neat scroll trigger animation.

Reviewer 2: ```suggestion
For this demo I have created a [Codepen](https://codepen.io/mrtrimble/pen/gOvLPgM??editors=0010) to follow along with. I already included HTML, CSS, GSAP, and the ScrollTrigger plugin, so no need to set anything up! We’ll focus on the JavaScript to find out that with a few lines of code, we can build a neat scroll trigger animation.
```

Tweens take in two key parameters:
- The element you want to animate.
- An object explaining how the element will animate, including CSS properties to animate.

You can include as many tweens as needed to complete the animation. GSAP includes three different types of tweens:
- `.to()` - animates an element *to* something.
- `.from()` - animates an element *from* something.
- `.fromTo()` - animates an element *from* one thing *to* another.

Pretty straightforward! To create tweens, we can tack them onto the timeline. To make things even easier, we can chain tweens together.

So in our example, we want to create a tween that scales the `.scale-text` element to a large size:

```js
const tl = new gsap.timeline();

tl.to('.scale-text', {
  scale: 10
});
```

The animation should trigger, however that is not quite what we want yet - we want this to happen on scroll. This is where the GSAP ScrollTrigger plugin comes in!

Inside where we initialized our timeline, we can configure ScrollTrigger:

```js
const tl = new gsap.timeline({
  scrollTrigger: {
    trigger: ".hero",

>     start: "top top",

Reviewer 2: I'd suggest using single quotes for this bit since you use single quotes above. Just kind of setting consistency

`scrub` - enables the scrollbar to become the scrubber for the animation. Setting this as `true` will enable this, but specifying a duration will smooth out an animation.

There is a lot of ways to configure ScrollTrigger. I recommend checking out the [GSAP ScrollTrigger documentaiton page](https://greensock.com/scrolltrigger/) to find out more!

Now, our animation may not look like it is doing anything, but when you start scrolling you should see it kick off. The text "empowering people" will scale up in size. Our scroll trigger animation is working and we did that with only a single tween too! 

Here are a couple of small enhancements we can add as well:

```js
const tl = new gsap.timeline({
  scrollTrigger: {
    trigger: ".hero",
    start: "top top",
    scrub: 1
  }
});

tl.to('.scale-text', {
  scale: 10
})
.to('.hero', {
  opacity: 0
}, "<25%")
.to('.cards', {
  opacity: 1,
  duration: 1,
  y: 0,
  stagger: 0.5,
  ease: "back.out(1.7)"
}, "<25%");
```

I added two more tweens:
- The first will fade out the `.hero` element as it scrolls out of view.
- The second will pop the cards into view as they scroll.

> This adds slightly more interesting effects to the page. There is a couple of other fun things to note happening as well: 

Reviewer 2: ```suggestion
This adds slightly more interesting effects to the page. There are a couple of other fun things to note happening as well: 
```

- Easing is a method of smoothing out animations and can make tweens feel more natural while transitioning. GSAP can handle easing with the `ease` property and has several built in easing functions. You can play with these eases with the [ease visualizer](https://greensock.com/ease-visualizer/) and find out which is best for your animations.
- GSAP also includes shorthand properties such as `x` and `y` that correspond with the CSS `transform` property. 

### Final Result
So there you have it, a super neat animated scroll trigger effect that is only a few lines of JavaScript!

<iframe height="500" style="width: 100%;" scrolling="no" title="Finished - Scroll Trigger Demo" src="https://codepen.io/mrtrimble/embed/zYRoGgZ/7545ac69b915f060e1585b886719a856?default-tab=result" frameborder="no" loading="lazy" allowtransparency="true" allowfullscreen="true">
  See the Pen <a href="https://codepen.io/mrtrimble/pen/zYRoGgZ/7545ac69b915f060e1585b886719a856">
  Finished - Scroll Trigger Demo</a> by Ryan Trimble (<a href="https://codepen.io/mrtrimble">@mrtrimble</a>)
  on <a href="https://codepen.io">CodePen</a>.
</iframe>

> (For best results, view on [Codepen](https://codepen.io/mrtrimble/pen/zYRoGgZ/7545ac69b915f060e1585b886719a856))

Reviewer: Nice addition
